ProjFS.Mac/PrjFSKext/KauthHandler.cpp

```diff
@@ -482,68 +492,89 @@ KEXT_STATIC int HandleFileOpOperation(
         
         bool isDirectory = (0 != vnode_isdir(currentVnode));
         
-        VirtualizationRootHandle root = RootHandle_None;
+        VirtualizationRootHandle toRoot = RootHandle_None, fromRoot = RootHandle_None;
         FsidInode vnodeFsidInode = {};
-        int pid;
-        if (!ShouldHandleFileOpEvent(
+        int pid = -1;
+        bool messageDestinationProvider = ShouldHandleFileOpEvent(
                 &perfTracer,
                 context,
                 currentVnode,
                 action,
                 isDirectory,
-                &root,
+                fromPath,
+                &toRoot,
+                &fromRoot,
                 &vnodeFsidInode,
-                &pid))
+                &pid);
+        bool messageSourceProvider = VirtualizationRoot_IsValidRootHandle(fromRoot);
+        
+        if (!messageDestinationProvider && !messageSourceProvider)
         {
             goto CleanupAndReturn;
         }
         
         char procname[MAXCOMLEN + 1];
         proc_name(pid, procname, MAXCOMLEN + 1);
 
+        MessageType messageType;
+        PrjFSPerfCounter perfCounter;
         if (KAUTH_FILEOP_RENAME == action)
         {
-            PerfSample renameSample(&perfTracer, PrjFSPerfCounter_FileOp_Renamed);
+            perfCounter = PrjFSPerfCounter_FileOp_Renamed;
             
-            MessageType messageType =
+            messageType =
                 isDirectory
                 ? MessageType_KtoU_NotifyDirectoryRenamed
                 : MessageType_KtoU_NotifyFileRenamed;
-
+        }
+        else
+        {
+            perfCounter = PrjFSPerfCounter_FileOp_HardLinkCreated;
+            
+            messageType = MessageType_KtoU_NotifyFileHardLinkCreated;
+        }
+        
+        if (messageDestinationProvider)
+        {
+            PerfSample sample(&perfTracer, perfCounter);
             int kauthResult;
-            int kauthError;
+            int kauthError = 0;
             if (!ProviderMessaging_TrySendRequestAndWaitForResponse(
-                    root,
+                    toRoot,
                     messageType,
                     currentVnode,
                     vnodeFsidInode,
                     newPath,
+                    fromPath,
                     pid,
                     procname,
                     &kauthResult,
                     &kauthError))
             {
-                goto CleanupAndReturn;
+                KextLog_Error("HandleFileOpOperation: Request %u to destination provider %d failed, kauthResult = %u, kauthError = %u",
```

Comment:
should this also CleanupAndReturn?  Same question below

Reply:
As far as I can tell, the only reason `ProviderMessaging_TrySendRequestAndWaitForResponse` would be returning false here is that the provider has gone offline since we checked that it was online, or that the provider itself returned an error message. Neither seems a good reason *not* to send a message to the source provider if that's what we were originally going to do.

Reply:
We generally call CleanupAndReturn which does a put if TrySendrequestAndWaitForResponse fails.  

My question is why would we not need it in this case.

If I'm missing something please let me know!

Reply:
There's no `return` prior to the `CleanupAndReturn:`, so we call `vnode_put(currentVnode)` if necessary in all call cases, so that should not have any bearing on the decision. So as far as I can tell the only difference with or without the `goto` is that without, we message the "from" provider even if messaging the "to" provider didn't succeed for whatever reason, assuming both must be messaged. Failure to message "to" isn't great, obviously, but I don't see how then *also* dropping the message to "from" would be helpful. (We can't fail the kauth call or anything like that, as FILEOP calls must always succeed.)